pkg/localmetrics/localmetrics.go

```diff
@@ -16,51 +16,167 @@ package localmetrics
 
 import (
 	"net/http"
+	"net/url"
+	"strings"
 	"time"
 
-	dms "github.com/openshift/deadmanssnitch-operator/pkg/controller/deadmanssnitch"
 	"github.com/prometheus/client_golang/prometheus"
 )
 
 var (
-	MetricDeadMansSnitchHeartbeat = prometheus.NewGaugeVec(prometheus.GaugeOpts{
-		Name: "metricDeadMansSnitchHeartbeat",
-		Help: "Metric for heartbeating the Dead Man's Snitch api",
-	}, []string{"name"})
+	Collector *MetricsCollector
+)
 
-	MetricsList = []prometheus.Collector{
-		MetricDeadMansSnitchHeartbeat,
-	}
+const (
+	operatorName      = "deadmanssnitch-operator"
+	snitchMethodLabel = "method"
 )
 
-// UpdateMetrics updates all the metrics ever 5 minutes
-func UpdateMetrics() {
+type MetricsCollector struct {
+	ReconcileDuration  prometheus.Histogram
+	apiCallDuration    *prometheus.HistogramVec
+	snitchCallErrors   prometheus.Counter
+	snitchCallDuration *prometheus.HistogramVec
+	collectors         []prometheus.Collector
+}
 
-	d := time.Tick(5 * time.Minute)
-	for range d {
-		UpdateMetricDeadMansSnitchHeartbeatGauge()
+func (m MetricsCollector) Describe(ch chan<- *prometheus.Desc) {
+	m.ReconcileDuration.Describe(ch)
+	m.apiCallDuration.Describe(ch)
+	m.snitchCallDuration.Describe(ch)
+	m.snitchCallErrors.Describe(ch)
+	for _, c := range m.collectors {
+		c.Describe(ch)
 	}
 }
 
-// UpdateMetricDeadMansSnitchHeartbeatGauge curls the DMS API, updates the gauge to 1 when successful.
-func UpdateMetricDeadMansSnitchHeartbeatGauge() {
+func (m MetricsCollector) Collect(ch chan<- prometheus.Metric) {
+	m.ReconcileDuration.Collect(ch)
+	m.apiCallDuration.Collect(ch)
+	m.snitchCallErrors.Collect(ch)
+	m.snitchCallDuration.Collect(ch)
+	for _, c := range m.collectors {
+		c.Collect(ch)
+	}
+}
 
-	req, err := http.NewRequest("GET", "https://api.deadmanssnitch.com/v1/snitches", nil)
-	if err != nil {
-		MetricDeadMansSnitchHeartbeat.With(prometheus.Labels{"name": "deadmanssnitch-operator"}).Set(float64(0))
+func NewMetricsCollector() *MetricsCollector {
+	return &MetricsCollector{
+		ReconcileDuration: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Name:        "dms_operator_reconcile_duration_seconds",
+			Help:        "The duration it takes to reconcile a ClusterDeployment",
+			ConstLabels: map[string]string{"name": operatorName},
+		}),
+		// apiCallDuration times API requests. Histogram also gives us a _count metric for free but also includes errors.
```

Comment:
Noticed after the fact (and I think this is my mistake, I think I flagged the wrong location in my previous comment): This isn't the right place for this comment, as apiCallDuration does *not* count errors. The comment needs to go next to `snitchCallDuration`.

I'm fixing this in the fup.

Reply:
Addressed in #69.